### PR TITLE
add server side Sentry

### PR DESCRIPTION
**Pull request by mathieud**
## Checklist before requesting a review
- [ ] I hereby declare my code ready for review.
- [ ] If it is a critical feature, I have added tests.
- [ ] The CI is successful.
- [ ] If there could backward compatibility issues, it has been discussed and planned.

### CODE REVIEW

File: packages/start-slicemachine/src/lib/sentrySetup.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/sentrySetup.ts]: `"@slicemachine/manager/*"` can be replaced with `"@slicemachine/manager". I'm not sure why `tsserver` uses the `/*` path when auto-importing. Any idea why?

File: packages/start-slicemachine/src/lib/sentrySetup.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/sentrySetup.ts]: Is there a reason we need to use a function? Could we determine this condition statically?

For example:

```
if (!process.env.NO_SENTRY) {
  return
}
```

---

If we plan to make `enableSentry()` return a value based on other conditions (e.g. a user's settings that may disable error tracking), we may want to convert that function into a method in the manager. `manager.telemetry.checkIsErrorTrackingAllowed()` could return a boolean.

---

`!process.env.NO_SENTRY` is a double negative, which gets negated again when we run `if (!enableSentry())`. It can be confusing to follow compared to a more explicit `SENTRY_ENABLED` constant. I can't find a better source right now, but [this article briefly describes why double negatives should be avoided](https://dev.to/mcsee/code-smell-51-double-negatives-4fh0).

---

`enableSentry` is a verb that makes me think I can call this function to enable Sentry. Something like `checkIsSentryEnabled()` more closely describes what the function does.

`checkIsSentryEnabled()` is preferred over `isSentryEnabled()` because the former lets you declare a variable named `isSentryEnabled`:

```ts
const isSentryEnabled = checkIsSentryEnabled()
```

File: packages/start-slicemachine/src/lib/sentrySetup.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/sentrySetup.ts]: This assignment is unnecessary since `userId` defaults to `undefined`.

File: packages/start-slicemachine/src/lib/sentrySetup.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/sentrySetup.ts]: I recognize that this is a subjective opinion, but using `user.shortId` throughout the code is better than saving it in a `userId` variable for a couple of reasons:

- `userId` is a different name than `shortId` and could miscommunicate what the variable holds. Consider that `user` contains a `userId` property, which is different from its `shortId`.
- `user.shortId` shows us exactly where the `shortId` value comes from: the `user` variable. In contrast, `userId` is disconnected from `user` and could theoretically contain any string.

File: packages/start-slicemachine/src/lib/sentrySetup.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/sentrySetup.ts]: `manager.getPrismicAuthManager()` is for low-level use when you need access to the whole `PrismicAuthManager` instance. In this case, we only need one of its methods (`getProfile()`), which is exposed via the `UserManager`.

This version should be used instead:

```ts
manager.user.getProfile()
```

(I know we're missing documentation on these methods. Sorry for the confusion!)

File: packages/start-slicemachine/src/lib/sentrySetup.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/sentrySetup.ts]: This URL is a constant that should live in a dedicated constants file. Because `start-slicemachine` is relative small, we could just put it in a new `src/constants.ts` file:

```ts
// src/constants.ts

export const SENTRY_EXPRESS_DSN = "https://b673ba8b041d4449a0fb0a38691882dd@o146123.ingest.sentry.io/4504179268845568";
```

File: packages/start-slicemachine/src/lib/sentrySetup.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/sentrySetup.ts]: `version` isn't clear about what the version is from, which could be confusing further down in the code. For example, `version` could mean SM version, Sentry version, or `start-slicemachine` version.

`const sliceMachineVersion` is a good alternative.

File: packages/start-slicemachine/src/lib/sentrySetup.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/sentrySetup.ts]: When happens if `process.env.NODE_ENV` is not defined? I'm not sure it would be defined in production since developers aren't specifying `NODE_ENV` in their shell.

File: packages/start-slicemachine/src/lib/sentrySetup.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/sentrySetup.ts]: This `onError()` function could live directly in `createSliceMachineServer()`. Moving it there more clearly shows the connection between an error thrown in `h3` and the data sent to Sentry. This is especially true since `onError()` is a generic name that does not convey that it is related to Sentry.

```ts
const app = createApp({
  onError: (error, event) => {
    if (args.manager.telemetry.checkIsErrorTrackingAllowed()) {
      // The rest of the Sentry code...
    }
  }
})
```

---

In general, files in `lib` should only export one function. The name of the file should match the name of the export so it is clear where to find functions throughout the project.

In this case, we should rename the file to `lib/initSentry.ts`.

File: packages/start-slicemachine/src/lib/createSliceMachineServer.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/createSliceMachineServer.ts]: Is this endpoint still needed? `"Test Error"` sounds like something used during development.

File: packages/start-slicemachine/src/lib/createSliceMachineServer.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/createSliceMachineServer.ts]: What is this endpoint used for? If we are catching errors and re-throwing this custom "Uncaught Error", are we losing metadata from the originally thrown error somewhere in the process?

**Review by angeloashmore — CHANGES_REQUESTED** [review]
Hey @mathieud, I left a bunch of comments through the review. Most of them are about naming, with others commenting about the locations of logic.

Let me know if you have any questions about what I posted! 🙂

File: packages/start-slicemachine/src/lib/createSliceMachineServer.ts
Comment (mathieud) [reply packages/start-slicemachine/src/lib/createSliceMachineServer.ts]: I used these urls to simulate a manager returning an error or throwing one. Will be deleted before we merge

File: packages/start-slicemachine/src/lib/sentrySetup.ts
Comment (mathieud) [reply packages/start-slicemachine/src/lib/sentrySetup.ts]: I made a change keeping all the Sentry code in its file (modifying the app's `onError` instead).
It allows to keep all the Sentry code in its file and not leak anywhere else in the app.

What do you think?

File: packages/start-slicemachine/src/lib/sentrySetup.ts
Comment (mathieud) [reply packages/start-slicemachine/src/lib/sentrySetup.ts]: Made some changes, let me know what you think.

Also I found [this](https://github.com/prismicio/slice-machine/blob/71f0fd828d77e43344970731cff0ce54ff81ee10/packages/manager/src/constants/API_TOKENS.ts#L8). Could I use it to check for production env? Or is based on the remote endpoint I will have false positives for prod?

File: packages/start-slicemachine/src/lib/sentrySetup.ts
Comment (mathieud) [reply packages/start-slicemachine/src/lib/sentrySetup.ts]: About the named export matching the filename, is it something we would want to automate? Found https://github.com/ekwoka/eslint-plugin-filename-export with a quick search, it looks like it does what we want. Should be relatively easy to adapt if we want a separate behaviour

File: packages/start-slicemachine/src/lib/setupSentry.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/setupSentry.ts]: Should this line be removed?

File: packages/start-slicemachine/src/lib/setupSentry.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/setupSentry.ts]: We're missing a type annotation here:

```ts
let user: PrismicUserProfile | undefined;
```

The type is exported from `@slicemachine/manager`:

```ts
import { PrismicUserProfile } from "@slicemachine/manager";
```

(Small nitpick, but since the value is the user's profile, not necessarily the "user", could we rename the variable to `profile` or `userProfile`?)

File: packages/start-slicemachine/src/lib/setupSentry.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/setupSentry.ts]: This could be slightly more readable if we wrapped it in an `if` statement instead of using optional chaining. The `if` statement more clearly states that the user is set only if `user` is populated. This is just my personal opinion, so feel free to ignore if you feel otherwise.

```ts
if (user) {
  Sentry.setUser({ id: user.shortId })
}
```

File: packages/start-slicemachine/src/lib/setupSentry.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/setupSentry.ts]: It looks like `isDevEnv()` is only used in `isSentryEnabled()`. We could clean up the code a bit and inline the development check:

```ts
["development", "test"].includes(process.env.NODE_ENV ?? "") || process.env.SENTRY_AUTH_TOKEN !== undefined;
```

Again, a stylistic choice, so I'll leave it up to you if you want to make this change. 🙂

File: packages/start-slicemachine/src/lib/setupSentry.ts
Comment (angeloashmore) [inline packages/start-slicemachine/src/lib/setupSentry.ts]: Could we check the error that is thrown instead? It should contain some metadata in the Error's `name` and `message` properties. Checking `error` would make this a more universal `onError` handler that also handles errors not thrown by r19.

`procedurePath` is an implementation detail of r19 that could change at any time. r19 does not expect developers to read the raw network request.

If the error does _not_ contain the metadata we need, let me know and we can try to make adjustments to r19 so it contains the metadata.

**Review by angeloashmore — COMMENTED** [review]
Hey Mathieu, the changes look good! I have a few new comments that should be easy to address. After those fixes, I think we're good to merge.

The r19/`procedurePath` TODO could be quick if the thrown `error` contains the metadata we need. If it doesn't, let me know and we can update the library. We're most likely going to change the `/_manager` endpoint to include the full procedure path in the URL, like: `/_manager/users/getProfile`, which would make it easy to log which URL threw an error.

File: packages/start-slicemachine/src/lib/setupSentry.ts
Comment (lihbr) [inline packages/start-slicemachine/src/lib/setupSentry.ts]: Maybe this could use `semver.prerelease` instead to know if any prerelease is present

File: packages/start-slicemachine/src/lib/setupSentry.ts
Comment (lihbr) [inline packages/start-slicemachine/src/lib/setupSentry.ts]: Maybe this logic could be done in the constant file (similar to https://github.com/prismicio/slice-machine/blob/71f0fd828d77e43344970731cff0ce54ff81ee10/packages/manager/src/constants/API_TOKENS.ts, but simpler), this way if this token is used elsewhere, then it's consistent

File: packages/start-slicemachine/src/lib/setupSentry.ts
Comment (lihbr) [inline packages/start-slicemachine/src/lib/setupSentry.ts]: Maybe we should also add an additional `enabled` getter as part of the `SliceMachineManager.telemetry` manager so that people preference on telemetry is also honored for Sentry

File: packages/start-slicemachine/src/lib/createSliceMachineServer.ts
Comment (lihbr) [reply packages/start-slicemachine/src/lib/createSliceMachineServer.ts]: I assume we can delete it now? @mathieud

File: packages/start-slicemachine/src/lib/createSliceMachineServer.ts
Comment (lihbr) [reply packages/start-slicemachine/src/lib/createSliceMachineServer.ts]: Same as above @mathieud?

**Review by lihbr — APPROVED** [review]

**Pull request merged**
